rfcs/text/0011_global_search.md

```diff
@@ -0,0 +1,693 @@
+
+## We could use plain string instead of an enum for `GlobalSearchResult.type`
+
+The current static enum used for type
+
+```ts
+enum SEARCH_TYPE {
+  // non-exhaustive
+  applications = 10,
+  dashboard = 20,
+  visualization = 30,
+  search = 40,
+}
+
+/** @public */
+type GlobalSearchResultType = keyof typeof SEARCH_TYPE;
+
+interface GlobalSearchResult {
+  // [...]
+  type: GlobalSearchResultType;
+}
+```
+
+has some limitations:
+- It forces the enum to be modified every time a new type is added
+- 3rd party plugins cannot introduce new types
+
+We could change the API to accept plain strings for `GlobalSearchResult.type`. However, atm this enum approach 
+is needed as the ordinal values of the entries is used in results sorting. Changing to plain strings forces to find 
+an alternative sorting logic.
+
+## triggered searches could be cancelable
+
+In current specifications, once a search has been triggered, it's not possible to cancel it (see [Search cancellation](#search-cancellation))
+
+Main drawback of this decision is that if a `find` API consumer is going to perform multiple `find` requests and 
+only uses the last one (For example in the UI, a debounced `find` triggered on end-user keypress: On every new keypress,
+a new search is performed and the previous, potentially still executing, `find` call will still be executed), it can't
+cancel the previous active search, resulting on potential useless http calls and resources consumption.
+
+We could add an optional cancellation signal or observable in `GlobalSearchOptions` and propagate it to result providers.
+That way `find` calls could be aborted.
+
+However result providers would still have to manually handles this signal to cancel any http call or 
+other asynchronous task that could be pending. 
+
+Notes:
+- As this can be implemented with an additional option, this would be a non-breaking change and so could be done at 
+  a later time (even if it means that existing result provider would not handles this option until they implements it).
+
+## The GlobalSearch service could be provided as a plugin instead of a core service
+
+The GlobalSearch API could be provided and exposed from a plugin instead of `core`.
+
+Pros:
+- Less `core` API exposure
+
+Cons:
+- We know our initial consumer of this API is going to use it from/for the chrome header component, and because this component 
+  is in `core`, that would mean creating a bridge of some kind to be able to use the service from core, as plugin APIs
+  are not usable here.
+- The platform team is going to provide the base result providers for SO and application search results, so having the GS
+  service in a plugin would mean creating yet another plugin for these providers instead of having them self contained
+  in core. 
+- We are probably going to be adding internal SO APIs for the SO result provider, as current APIs are not sufficient to
```

Comment:
I think it's worth fleshing out how such an SO result provider might work. Simply searching across all fields in all saved objects types might be the easiest first step, but results might not be optimal. I'm also unsure about how we would render results, if we got a hit in a `panelsJSON` field do we render the JSON structure or just the title of the saved object (which wouldn't include any relevant keywords)?

Letting each plugin register a saved object result provider would be a lot more work, but it's probably at least worth mentioning as an alternative. This would allow plugins to tweak the priorities of their fields, but also provide a way to render a search hit with human readable context like "visualization field name: *your hit*" instead of "panelsJSON.vizualizations[0].fields[3].title: *your hit*" (just a made up example, not a real path).

Reply:
The SO result provider(s?) is another (very) complex topic on its own, but is outside of the scope of this RFC, that only defines the GS API.

Unless you see reasons not to, I would like to separate the two subjects, as they are both big enough on their own.

Reply:
I agree we can flesh this out in a separate RFC, however, if the internal SO API's become the biggest reason for putting this into Core, I think we should at least have some idea for what these API's will do and how they'll work (even if it's a separate draft RFC or whatever).